cpp/src/file_handle.cpp

```diff
@@ -23,13 +26,77 @@
 
 namespace kvikio {
 
+namespace {
+
+/**
+ * @brief Get a thread pool specific to the block device hosting the given file.
+ *
+ * Thread pools are created lazily on first access and cached for subsequent lookups.
+ * Two levels of caching are used:
+ * - File path --> thread pool (fast path for repeated access to the same file)
+ * - Block device ID --> thread pool (groups different files on the same device)
+ *
+ * @param file_path Path to the file where a thread pool is requested for the underlying block
+ * device.
+ * @return Pointer to the appropriate thread pool. The pointer remains valid for the lifetime of the
+ * program (static storage duration).
+ *
+ * @note If device detection fails for any reason (e.g., unsupported filesystem, permission issues),
+ * the default global thread pool is returned and an error is logged.
+ */
+ThreadPool* get_thread_pool_per_block_device(std::string const& file_path)
+{
+  KVIKIO_NVTX_FUNC_RANGE();
+
+  if (!defaults::thread_pool_per_block_device()) { return &defaults::thread_pool(); }
+
+  static std::mutex mtx;
+  static std::unordered_map<std::string, std::shared_ptr<ThreadPool>> file_path_to_thread_pool_map;
+  static std::unordered_map<dev_t, std::shared_ptr<ThreadPool>> dev_id_to_thread_pool_map;
+
+  try {
+    // Fast path: check if this exact file path has been seen before
+    {
+      std::lock_guard lock(mtx);
+      if (auto it = file_path_to_thread_pool_map.find(file_path);
+          it != file_path_to_thread_pool_map.end()) {
+        return it->second.get();
+      }
+    }
+
+    // Resolve file path to its underlying block device
+    auto block_dev_info = get_block_device_info(file_path);
+
+    // Check if we already have a thread pool for this block device
+    std::lock_guard lock(mtx);
+    if (auto it = dev_id_to_thread_pool_map.find(block_dev_info.id);
+        it != dev_id_to_thread_pool_map.end()) {
+      // Cache the file path mapping for future fast-path lookups
```

Comment:
would it be possible to cache using a parent directory, so the result applies to multiple files? I'm really not sure what's feasible here, just an idea.